src/utils/upload/FileUploader.js

```diff
@@ -0,0 +1,175 @@
+import _ from 'lodash';
+
+import { AsyncGzip } from 'fflate';
+import filereaderStream from 'filereader-stream';
+
+import putPartInS3 from './putPartInS3';
+import UploadStatus from './UploadStatus';
+
+class FileUploader {
+  constructor(
+    file,
+    compress,
+    chunkSize,
+    uploadParams,
+    abortController,
+    onStatusUpdate,
+  ) {
+    if (!file
+      || !chunkSize
+      || !uploadParams
+      || !abortController
+      || !onStatusUpdate
+    ) {
+      throw new Error('FileUploader: Missing required parameters');
+    }
+
+    this.file = file;
+    this.compress = compress;
+    this.chunkSize = chunkSize;
+    this.uploadParams = uploadParams;
+
+    // Upload related callbacks and handling
+    this.onStatusUpdate = onStatusUpdate;
+    this.abortController = abortController;
+
+    // Stream handling
+    this.totalChunks = Math.ceil(file.size / chunkSize);
+    this.pendingChunks = this.totalChunks;
+
+    // Used to assign partNumbers to each chunk
+    this.partNumberIt = 0;
+
+    this.readStream = null;
+    this.gzipStream = null;
+
+    this.uploadedParts = [];
+
+    this.resolve = null;
+    this.reject = null;
+
+    // To track upload progress
+    this.uploadedPartPercentages = new Array(this.totalChunks).fill(0);
+
+    this.currentChunk = null;
+  }
+
+  async upload() {
+    return new Promise((resolve, reject) => {
+      this.resolve = resolve;
+      this.reject = reject;
+
+      this.readStream = filereaderStream(
+        this.file?.fileObject || this.file, { chunkSize: this.chunkSize },
+      );
+
+      this.#setupReadStreamHandlers();
+
+      if (this.compress) {
+        this.gzipStream = new AsyncGzip({ level: 1, consume: false });
+        this.#setupGzipStreamHandlers();
+      }
+    });
+  }
+
+  #uploadChunk = async (compressedPart, partNumber) => {
```

Comment:
what does this `#` in front of the function name mean? I see it for first time

Reply:
yes, it's a private thing